sshd-core/src/main/java/org/apache/sshd/common/io/nio2/Nio2Session.java

```diff
@@ -382,7 +385,36 @@ protected void handleReadCycleFailure(ByteBuffer buffer, Readable bufReader, Thr
         exceptionCaught(exc);
     }
 
+    @Override
+    public void suspendRead() {
+        log.debug("suspendRead({})", this);
+        suspendRead.set(true);
+    }
+
+    @Override
+    public void resumeRead() {
+        Runnable runnable = readRunnable.getAndSet(null);
+        if (runnable != null) {
+            log.debug("resumeRead({})", this);
+            suspendRead.set(false);
```

Comment:
Shouldn't this be outside the "if", between lines 396/397?

Reply:
I don't think this will change the behavior, but it will be cleaner.